### PR TITLE
Quantum debugging (feature branch)

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
Feature branch for the new quantum debugging feature

### CODE REVIEW

**Review by soranjh — APPROVED** [review]
Thanks @Jaybsoni, ready to be merged.

File: pennylane/debugging.py [hunk 1]
```diff
@@ -105,3 +110,198 @@ def get_snapshots(*args, **kwargs):
         return dbg.snapshots
 
     return get_snapshots
+
+
+class PLDB(pdb.Pdb):
+    """Custom debugging class integrated with Pdb.
```
Comment (obliviateandsurrender) [inline pennylane/debugging.py:116]: Might be nice to link "Pdb" with the python debugger.

File: pennylane/debugging.py [hunk 2]
```diff
@@ -105,3 +110,198 @@ def get_snapshots(*args, **kwargs):
...
+        new_batch, fn = program(valid_batch)
+
+        # TODO: remove [0] index once compatible with transforms
+        return fn(dev.execute(new_batch, new_config))[0]
+
+
+@contextmanager
+def pldb_device_manager(device):
+    """Context manager to automatically set and reset active
+    device on the PLDB debugger.
```
Comment (obliviateandsurrender) [inline pennylane/debugging.py:211]:
```suggestion
    device on the Pennylane Debugger (PLDB).
```

**Review by obliviateandsurrender — APPROVED** [review]
Thanks @Jaybsoni, it looks good to me! A couple of optional suggestions; otherwise just missing a `changelog` entry.

File: pennylane/debugging.py (hunk 1 above)
Comment (Jaybsoni) [reply pennylane/debugging.py:116]: This will be addressed in the docs PR 👍🏼

**Pull request merged**